### PR TITLE
Feature/add product

### PR DESCRIPTION
**Pull request by NadaFeteiha**

### CODE REVIEW

File: src/main/kotlin/com/the_chance/data/models/Product.kt [hunk 1]
```diff
@@ -4,7 +4,9 @@ import kotlinx.serialization.Serializable
 
 @Serializable
 data class Product(
-    val id: String,
+    val id: Long,
     val name: String,
+    val quantity: String?,
     val price: Double,
+    val isDeleted: Boolean = false
```
Comment (ahmednasserzaza) [inline src/main/kotlin/com/the_chance/data/models/Product.kt]: I think isDeleted is not nesessary here

File: src/main/kotlin/com/the_chance/data/services/ProductSchema.kt [hunk 2]
```diff
@@ -2,39 +2,67 @@ package com.the_chance.data.services
 
 import com.the_chance.data.models.Product
 import com.the_chance.data.tables.ProductTable
-import org.jetbrains.exposed.sql.transactions.transaction
-import org.jetbrains.exposed.sql.transactions.experimental.newSuspendedTransaction
-import kotlinx.coroutines.Dispatchers
 import org.jetbrains.exposed.sql.*
+import org.jetbrains.exposed.sql.transactions.transaction
 
 class ProductService(private val database: Database) : BaseService() {
```
Comment (ahmednasserzaza) [inline src/main/kotlin/com/the_chance/data/services/ProductSchema.kt]: You can remove (private val) in constructor , you can access it directrly in init block

File: src/main/kotlin/com/the_chance/data/services/ProductSchema.kt [hunk 3]
```diff
@@ -2,39 +2,67 @@ package com.the_chance.data.services
...
 
-    suspend fun create(productName: String, productPrice: Double): Product = dbQuery {
-        val newProduct = ProductTable.insert {
-            it[name] = productName
-            it[price] = productPrice
+    suspend fun create(productName: String, productPrice: Double, productQuantity: String): Product = dbQuery {
+        val newProduct = ProductTable.insert { productRow ->
+            productRow[name] = productName
+            productRow[price] = productPrice
+            productRow[isDeleted] = false
```
Comment (ahmednasserzaza) [inline src/main/kotlin/com/the_chance/data/services/ProductSchema.kt]: It's not necessary here to insert isDeleted , In ProductTable you can make default value to this column as false

File: src/main/kotlin/com/the_chance/data/services/ProductSchema.kt [hunk 4]
```diff
@@ -2,39 +2,67 @@ package com.the_chance.data.services
...
     }
+
+    suspend fun updateProduct(
+        productId: Long,
+        productName: String?,
+        productPrice: Double?,
+        productQuantity: String?
+    ): Int {
+        return dbQuery {
+            ProductTable.update({ ProductTable.id eq productId }) { productRow ->
```
Comment (ahmednasserzaza) [inline src/main/kotlin/com/the_chance/data/services/ProductSchema.kt]: You can use singleOrNull function here 😄

File: src/main/kotlin/com/the_chance/data/tables/ProductTable.kt [hunk 5]
```diff
@@ -1,8 +1,10 @@
 package com.the_chance.data.tables
 
-import org.jetbrains.exposed.dao.id.UUIDTable
+import org.jetbrains.exposed.dao.id.LongIdTable
 
-object ProductTable : UUIDTable() {
+object ProductTable : LongIdTable() {
```
Comment (ahmednasserzaza) [inline src/main/kotlin/com/the_chance/data/tables/ProductTable.kt:5]: I fogot to change mine to LongIdTable 😆

File: src/main/kotlin/com/the_chance/data/tables/ProductTable.kt [hunk 6]
```diff
@@ -1,8 +1,10 @@
...
     val name = text("name")
     val price = double("price")
+    val quantity = text("quantity")
+    val isDeleted = bool("isDeleted")
```
Comment (ahmednasserzaza) [inline src/main/kotlin/com/the_chance/data/tables/ProductTable.kt]: Yes, Here in this line you can make it take default value as false So, You don't need to initiate it when you create product.
like that ===> 
val isDeleted = bool("isDeleted").default(false)

File: src/main/kotlin/com/the_chance/endpoints/Products.kt [hunk 7]
```diff
@@ -9,23 +9,67 @@ import io.ktor.server.request.*
...
 
-    get("/products"){
-        val products = productService.getAllProducts()
-        call.respond(ServerResponse.success(products))
-    }
+    route("/products") {
+
+        get {
+            val products = productService.getAllProducts()
+            call.respond(ServerResponse.success(products))
```
Comment (ahmednasserzaza) [inline src/main/kotlin/com/the_chance/endpoints/Products.kt:20]: Isn't necessary to put HttpStatusCode ?!

File: src/main/kotlin/com/the_chance/endpoints/Products.kt [hunk 8]
```diff
@@ -9,23 +9,67 @@ import io.ktor.server.request.*
...
+            val productPrice = params["price"]?.trim()?.toDoubleOrNull().orZero()
+            val productQuantity = params["quantity"]?.trim().orEmpty()
+
+            if (productName.length < 4) {
+                call.respond(ServerResponse.error("Product name length should be 4 characters or more"))
+            } else if (productPrice < 0.0) {
+                call.respond(ServerResponse.error("Product Price should be positive number."))
+            } else {
+                val newAddedProduct = productService.create(productName, productPrice, productQuantity)
+                call.respond(HttpStatusCode.Created, ServerResponse.success(newAddedProduct))
```
Comment (ahmednasserzaza) [inline src/main/kotlin/com/the_chance/endpoints/Products.kt:31]: Isn't good to show message that product is added successfully ?!
I won't be happy if i made a request and get a success message = null

File: src/main/kotlin/com/the_chance/endpoints/Products.kt [hunk 9]
```diff
@@ -9,23 +9,67 @@ import io.ktor.server.request.*
...
+            }
+        }
+
+        put("{id}") {
+            val productId = call.parameters["id"]?.trim()?.toLongOrNull() ?: 0L
```
Comment (ahmednasserzaza) [inline src/main/kotlin/com/the_chance/endpoints/Products.kt]: You don't need to use elvis operator , just .toLongOrNull is enough

File: src/main/kotlin/com/the_chance/endpoints/Products.kt [hunk 10]
```diff
@@ -9,23 +9,67 @@ import io.ktor.server.request.*
...
+
+            val params = call.receiveParameters()
+            val productName = params["name"]?.trim().orEmpty()
+            val productPrice = params["price"]?.trim()?.toDoubleOrNull().orZero()
+            val productQuantity = params["quantity"]?.trim().orEmpty()
+
+            if (productId < 1) {
```
Comment (ahmednasserzaza) [inline src/main/kotlin/com/the_chance/endpoints/Products.kt]: you can check nullability here insted of this 
like that ==>
if ( productId == null ) ===> Show message "Inavalid product Id"

File: src/main/kotlin/com/the_chance/endpoints/Products.kt [hunk 11]
```diff
@@ -9,23 +9,67 @@ import io.ktor.server.request.*
...
+                call.respond(ServerResponse.error("Not valid Product Id"))
+            } else if (!productService.isValid(
+                    productName = productName,
+                    productPrice = productPrice,
+                    productQuantity = productQuantity
+                )
+            ) {
+                call.respond(ServerResponse.error("Not valid Request."))
+            } else {
```
Comment (ahmednasserzaza) [inline src/main/kotlin/com/the_chance/endpoints/Products.kt]: You can make try and catch here if you will use singleOrNull and print the exception message

File: src/main/kotlin/com/the_chance/endpoints/Products.kt [hunk 12]
```diff
@@ -9,23 +9,67 @@ import io.ktor.server.request.*
...
+            }
+        }
 
-        if (productName.length < 4){
-            call.respond(ServerResponse.error("Product name length should be 4 characters or more"))
-        } else {
-            val newAddedProduct = productService.create(productName, productPrice)
-            call.respond(HttpStatusCode.Created, ServerResponse.success(newAddedProduct))
+        delete("{id}") {
+            val productId = call.parameters["id"]?.trim()?.toLongOrNull() ?: 0L
```
Comment (ahmednasserzaza) [inline src/main/kotlin/com/the_chance/endpoints/Products.kt]: The same as update

**Review by ahmednasserzaza — APPROVED** [review]
Great Work

File: src/main/kotlin/com/the_chance/endpoints/Products.kt (hunk 9 above)
Comment (NadaFeteiha) [reply src/main/kotlin/com/the_chance/endpoints/Products.kt]: @ahmednasserzaza no, I don't want productId to be null so I need to be 0 if null.

**Pull request merged**